### PR TITLE
Add Bitbucket support

### PR DESCRIPTION
**Pull request by mgiaccone**
Hi! I added support for Bitbucket. 

Unfortunately, the access to a private repository via HTTPS is not supported as Bitbucket will ask for a password and putting it in clear in the configuration would not be a great idea.

### CODE REVIEW

**Comment by abiosoft** [comment]
Thanks a lot for this.

The HTTPS password restriction is not limited to BitBucket and I experienced it with GitHub also. What I did was to convert the url to ssh scheme at runtime.

File: bitbucket_hook.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+var bitbucketIPBlocks = []string{
+	"131.103.20.160/27",
+	"165.254.145.0/26",
+	"104.192.143.0/24",
+}
+
+type BitbucketHook struct {}
+
+type bbPush struct {
+	Push struct {
```
Comment (abiosoft) [inline bitbucket_hook.go]: Though this is minor, it does not look like `gofmt` is executed for this source.

File: bitbucket_hook.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	err := json.Unmarshal(body, &push)
+	if err != nil {
+		return err
+	}
+
+	if len(push.Push.Changes) == 0 {
+		return errors.New("the push was incomplete, missing change list")
+	}
+
+    change := push.Push.Changes[0]
```
Comment (abiosoft) [inline bitbucket_hook.go]: `gofmt`

**Comment by mgiaccone** [comment]
Sorry, forgot about gofmt. Fixed it and removed the warning about bitbucket HTTPS in setup.go

**Comment by abiosoft** [comment]
Looks good. 

One last thing, I know this sounds harsh. I'd appreciate if you rebase the commits. 

Thanks :+1:

**Comment by mgiaccone** [comment]
Great. This will sound odd ;-) 
I never had to use rebase before, if you could tell me how I will be immensely grateful

**Comment by abiosoft** [comment]
Alright. This three steps should do. You can backup your source for precaution.

Assuming this repo is at `upstream` and your fork is at `origin`

1). Rebase
```
git rebase -i upstream/master
```
2). You'll get something like this in vim. Replace first commit message if required. Then change `pick` to `fixup` for all commits except the first. In this case `a68cbab` up to `6b42631`. Save and quit.
```
pick b5e19cc first commit message
pick a68cbab 2nd commit message
pick ddb13a1 3rd commit message
pick 6b42631 another commit message
# Rebase de8b3e2..d7f3632 onto de8b3e2
#
# Commands:
#  p, pick = use commit
#  r, reword = use commit, but edit the commit message
#  e, edit = use commit, but stop for amending
#  s, squash = use commit, but meld into previous commit
#  f, fixup = like "squash", but discard this commit's log message
#  x, exec = run command (the rest of the line) using shell
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
# Note that empty commits are commented out
```
3). Do a force push to your fork.
```
git push -f origin master
```

That is all and this PR will be updated with the all 11 commits rebased to 1.

**Comment by mgiaccone** [comment]
Thanks for the explanation. I ended up doing that in another branch to avoid forcing the push :-)

**Comment by abiosoft** [comment]
Perfectly fine by me. I hope you learned from this.

Thanks for the effort.

**Comment by mgiaccone** [comment]
Sure, 16 years as a developer on my shoulders and still learning every day. Love it!
Thank you

**Comment by abiosoft** [comment]
Wow, that's a big chunk of my age. :blush: 

I equally learned from your contribution.

**Pull request closed without merging**